README.md
jcoolib
=======

A Cartesian coordinate system library for Java Swing


NEWS
Thu Aug 21, 2014 - 0.1.2
===========================
* BUG FIX: Numbers on the unit lines are now visible when the lines are close to
           the edge of the screen.
* BUG FIX: Fixed a bug that caused lines going from the bottom right corner to
           the upper left corner not to be drawn in some cases.

Fri Aug 15, 2014 - 0.1.1
===========================
* BUG FIX: Fixed a bug that caused lines going from the bottom left corner to
           the upper right corner not to be drawn in some cases.

Mon Jul 29, 2013 - 0.1.0
===========================
Initial release.


src/test/Main.java
package test;

import java.awt.Color;
import java.util.Random;

import javax.swing.JFrame;

import cartesian.coordinate.CCPoint;
import cartesian.coordinate.CCPolygon;
import cartesian.coordinate.CCSystem;
import cartesian.coordinate.CCLine;

public class Main extends JFrame {
    private static final long serialVersionUID = 1L;
    
    Main() {
        super("Viewer");
        setTitle("Viewer");
        
        setVisible(true);
        setSize(800, 600);
        setLocationRelativeTo(null);
        CCSystem s = new CCSystem();
        add(s);
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        
        
        double[] x = new double[]{-1,2,0};
        double[] y = new double[]{-1,-2,3};
        CCPolygon ccp = new CCPolygon(x, y);
        
        s.add(new CCLine(4.0, 1.0, Color.blue));
        s.add(new CCPoint(4, 5));
        s.add(ccp);
        
        /*
        Random r = new Random();
        for (int i = 0; i < 300; i++) {
            double a = r.nextDouble();
            double b = r.nextDouble();
            double c = r.nextDouble();
            if (r.nextBoolean()) a = -a;
            if (r.nextBoolean()) b = -b;
            if (r.nextBoolean()) c = -c;
            
            int cint = r.nextInt(13);
            Color color;
            
            switch (cint) {
            case 0: color = Color.black; break;
            case 1: color = Color.blue; break;
            case 2: color = Color.cyan; break;
            case 3: color = Color.darkGray; break;
            case 4: color = Color.gray; break;
            case 5: color = Color.green; break;
            case 6: color = Color.lightGray; break;
            case 7: color = Color.magenta; break;
            case 8: color = Color.orange; break;
            case 9: color = Color.pink; break;
            case 10: color = Color.red; break;
            case 11: color = Color.white; break;
            default: color = Color.yellow;
            }
            
            s.add(new CCLine(a,b,c,color));
        }
        */
    }
    public static void main(String[] args) {
        new Main();
    }
}


src/cartesian/coordinate/Line.java
/*
 * Copyright (C) 2012 Andreas Halle
 *
 * This file is part of jcoolib
 *
 * jcoolib is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * jcoolib is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public license
 * along with jcoolib. If not, see <http://www.gnu.org/licenses/>.
 */
package cartesian.coordinate;

import java.awt.BasicStroke;
import java.awt.Color;
import java.awt.Paint;
import java.awt.Stroke;

/**
 * {@code Line} represents a straight line in a Cartesian coordinate system.
 * <p>
 * Lines are stored as a linear equation on the format:
 * <pre>
 *     ax + by = c
 * </pre>
 *
 * @author Andreas Halle
 * @see    CCSystem
 */
public class Line {
    protected double a;
    protected double b;
    protected double c;
    protected Paint paint;
    protected Stroke stroke;
    
    
    
    /**
     * Create a new line from a linear equation in slope-intercept form:
     * <pre>
     *     y = mx + b
     * </pre>
     * 
     * @param m
     *        The slope of the line.
     * @param b
     *        y-intercept; the y-coordinate of the location where the line
     *        crosses the y axis.
     */
    public Line(double m, double b) {
        this(-m, 1, b);
    }



    /**
     * Create a new line from a linear equation in general form:
     * <pre>
     *     ax + by = c
     * </pre>
     * 
     * @param a
     *        Coefficient of the x-variable.
     * @param b
     *        Coefficient of the y-variable.
     * @param c
     *        A constant.
     */
    public Line(double a, double b, double c) {
        this(a, b, c, Color.black, new BasicStroke(1f));
    }
    
    
    
    /**
     * Create a new line from a linear equation in general form:
     * <pre>
     *     ax + by = c
     * </pre>
     * 
     * @param a
     *        Coefficient of the x-variable.
     * @param b
     *        Coefficient of the y-variable.
     * @param c
     *        A constant.
     * @param paint
     *        Draw the line using this paint.
     */
    public Line(double a, double b, double c, Paint paint) {
        this(a, b, c, paint, new BasicStroke(1f));
    }
    
    
    
    /**
     * Create a new line from a linear equation in slope-intercept form:
     * <pre>
     *     y = mx + b
     * </pre>
     * 
     * @param m
     *        The slope of the line.
     * @param b
     *        y-intercept; the y-coordinate of the location where the line
     *        crosses the y axis.
     * @param paint
     *        Draw the line using this paint.
     */
    public Line(double m, double b, Paint paint) {
        this(-m, 1, b, paint);
    }



    /**
     * Create a new line from a linear equation in general form:
     * <pre>
     *     ax + by = c
     * </pre>
     * 
     * @param a
     *        Coefficient of the x-variable.
     * @param b
     *        Coefficient of the y-variable.
     * @param c
     *        A constant.
     * @param paint
     *        Draw the line using this paint.
     * @param stroke
     *        Draw the line using this stroke.
     */
    public Line(double a, double b, double c, Paint paint, Stroke stroke) {
        this.a = a;
        this.b = b;
        this.c = c;
        this.paint = paint;
        this.stroke = stroke;
    }
    
    
    
    /**
     * Create a new line from a linear equation in slope-intercept form:
     * <pre>
     *     y = mx + b
     * </pre>
     * 
     * @param m
     *        The slope of the line.
     * @param b
     *        y-intercept; the y-coordinate of the location where the line
     *        crosses the y axis.
     * @param paint
     *        Draw the line using this paint.
     * @param stroke
     *        Draw the line using this stroke.
     */
    public Line(double m, double b, Paint paint, Stroke stroke) {
        this(-m, 1, b, paint, stroke);
    }
    
    
    
    /**
     * Find the x-value of the line at a specific y-value.
     * 
     * @param  y
     *         y value of the function.
     * @return
     *         x value at the given y.
     */
    public double solveForX(double y) {
        return (c-b*y)/a;
    }
    
    
    
    /**
     * Find the y-value of the line at a specific x-value.
     * 
     * @param  x
     *         x value of the function.
     * @return
     *         y value at the given x.
     */
    public double solveForY(double x) {
        return (c-a*x)/b;
    }
}

src/cartesian/coordinate/CCLine.java
/*
 * Copyright (C) 2012 Andreas Halle
 *
 * This file is part of jcoolib
 *
 * jcoolib is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * jcoolib is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public license
 * along with jcoolib. If not, see <http://www.gnu.org/licenses/>.
 */
package cartesian.coordinate;

import java.awt.BasicStroke;
import java.awt.Color;
import java.awt.Paint;
import java.awt.Stroke;

/**
 * {@code CCLine} represents a straight line in a Cartesian coordinate system.
 * <p>
 * Lines are stored as a linear equation on the format:
 * <pre>
 *     ax + by = c
 * </pre>
 *
 * @author Andreas Halle
 * @see    CCSystem
 */
public class CCLine {
    protected double a;
    protected double b;
    protected double c;
    protected Paint paint;
    protected Stroke stroke;
    
    
    
    /**
     * Create a new line from a linear equation in slope-intercept form:
     * <pre>
     *     y = mx + b
     * </pre>
     * 
     * @param m
     *        The slope of the line.
     * @param b
     *        y-intercept; the y-coordinate of the location where the line
     *        crosses the y axis.
     */
    public CCLine(double m, double b) {
        this(-m, 1, b);
    }



    /**
     * Create a new line from a linear equation in general form:
     * <pre>
     *     ax + by = c
     * </pre>
     * 
     * @param a
     *        Coefficient of the x-variable.
     * @param b
     *        Coefficient of the y-variable.
     * @param c
     *        A constant.
     */
    public CCLine(double a, double b, double c) {
        this(a, b, c, Color.black, new BasicStroke(1f));
    }
    
    
    
    /**
     * Create a new line from a linear equation in general form:
     * <pre>
     *     ax + by = c
     * </pre>
     * 
     * @param a
     *        Coefficient of the x-variable.
     * @param b
     *        Coefficient of the y-variable.
     * @param c
     *        A constant.
     * @param paint
     *        Draw the line using this paint.
     */
    public CCLine(double a, double b, double c, Paint paint) {
        this(a, b, c, paint, new BasicStroke(1f));
    }
    
    
    
    /**
     * Create a new line from a linear equation in slope-intercept form:
     * <pre>
     *     y = mx + b
     * </pre>
     * 
     * @param m
     *        The slope of the line.
     * @param b
     *        y-intercept; the y-coordinate of the location where the line
     *        crosses the y axis.
     * @param paint
     *        Draw the line using this paint.
     */
    public CCLine(double m, double b, Paint paint) {
        this(-m, 1, b, paint);
    }



    /**
     * Create a new line from a linear equation in general form:
     * <pre>
     *     ax + by = c
     * </pre>
     * 
     * @param a
     *        Coefficient of the x-variable.
     * @param b
     *        Coefficient of the y-variable.
     * @param c
     *        A constant.
     * @param paint
     *        Draw the line using this paint.
     * @param stroke
     *        Draw the line using this stroke.
     */
    public CCLine(double a, double b, double c, Paint paint, Stroke stroke) {
        this.a = a;
        this.b = b;
        this.c = c;
        this.paint = paint;
        this.stroke = stroke;
    }
    
    
    
    /**
     * Create a new line from a linear equation in slope-intercept form:
     * <pre>
     *     y = mx + b
     * </pre>
     * 
     * @param m
     *        The slope of the line.
     * @param b
     *        y-intercept; the y-coordinate of the location where the line
     *        crosses the y axis.
     * @param paint
     *        Draw the line using this paint.
     * @param stroke
     *        Draw the line using this stroke.
     */
    public CCLine(double m, double b, Paint paint, Stroke stroke) {
        this(-m, 1, b, paint, stroke);
    }
    
    
    
    /**
     * Find the x-value of the line at a specific y-value.
     * 
     * @param  y
     *         y value of the function.
     * @return
     *         x value at the given y.
     */
    public double solveForX(double y) {
        return (c-b*y)/a;
    }
    
    
    
    /**
     * Find the y-value of the line at a specific x-value.
     * 
     * @param  x
     *         x value of the function.
     * @return
     *         y value at the given x.
     */
    public double solveForY(double x) {
        return (c-a*x)/b;
    }
}

src/cartesian/coordinate/CCPoint.java
/*
 * Copyright (C) 2012 Andreas Halle
 *
 * This file is part of jcoolib
 *
 * jcoolib is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * jcoolib is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public license
 * along with jcoolib. If not, see <http://www.gnu.org/licenses/>.
 */
package cartesian.coordinate;

import java.awt.BasicStroke;
import java.awt.Color;
import java.awt.Paint;
import java.awt.Stroke;

/**
 * {@code CCPoint} represents a point in a Cartesian coordinate system.
 *
 * @author Andreas Halle
 * @see    CCSystem
 */
public class CCPoint {
    protected double x;
    protected double y;
    protected Paint paint;
    protected Stroke stroke;
    
    
    
    /**
     * Create a point at given coordinates.
     * 
     * @param x
     *        x-coordinate for the location of the point.
     * @param y
     *        y-coordinate for the location of the point.
     * @param paint
     *        {@code Paint} to use when painting a circle at this point's
     *        location.
     * @param stroke
     *        Draw the circle at this point's location using this {@code Stroke}
     *        
     */
    public CCPoint(double x, double y, Paint paint, Stroke stroke) {
        this.x = x;
        this.y = y;
        this.paint = paint;
        this.stroke = stroke;
    }
    
    
    
    /**
     * Create a point at given coordinates.
     * <p>
     * The point will be painted in black with a 1 pixel thick edge.
     * 
     * @param x
     *        x-coordinate for the location of the point.
     * @param y
     *        y-coordinate for the location of the point.
     */
    public CCPoint(double x, double y) {
        this(x, y, Color.black, new BasicStroke(1f));
    }
}

src/cartesian/coordinate/CCPolygon.java
/*
 * Copyright (C) 2012 Andreas Halle
 *
 * This file is part of jcoolib
 *
 * jcoolib is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * jcoolib is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public license
 * along with jcoolib. If not, see <http://www.gnu.org/licenses/>.
 */
package cartesian.coordinate;

import java.awt.BasicStroke;
import java.awt.Color;
import java.awt.Paint;
import java.awt.Stroke;

/**
 * {@code CCPolygon} represents a polygon in a Cartesian coordinate system, i.e.
 * a shape consisting of straight lines joined through points to form a circuit.
 *
 * @author Andreas Halle
 * @see    CCSystem
 */
public class CCPolygon {
    protected double[] xpoints;
    protected double[] ypoints;
    protected Paint paint;
    protected Paint fill;
    protected Stroke stroke;
    
    
    
    /**
     * Create a polygon from a set of points. A polygon needs more than two
     * points.
     * 
     * @param xpoints
     *        x-coordinates for the points that form the polygon.
     * @param ypoints
     *        y-coordinates for the points that form the polygon.
     * @param paint
     *        {@code Paint} to paint the edges of the polygon with.
     * @param fill
     *        {@code Paint} to fill the interior of the polygon with.
     * @param stroke
     *        Draw the edges of the polygon with this {@code Stroke}.
     */
    public CCPolygon(double [] xpoints, double[] ypoints, Paint paint,
                                            Paint fill, Stroke stroke) {
        if (xpoints.length != ypoints.length) {
            String e = "number of x-coordinates must match number of y"
                     + "-coordinates";
            throw new IllegalArgumentException(e);
        }
        if (xpoints.length < 2) {
            String e = "cannot create a polygon from less than three points.";
            throw new IllegalArgumentException(e);
        }
        
        this.xpoints = xpoints;
        this.ypoints = ypoints;
        this.paint = paint;
        this.fill = fill;
        this.stroke = stroke;
    }
    
    
    
    /**
     * Create a polygon from a set of points. A polygon needs more than two
     * points.
     * <p>
     * The edges will be painted in black with a 1 pixel thick edge. The
     * interior of the polygon will be filled in pink.
     * 
     * @param xpoints
     *        x-coordinates for the points that form the polygon.
     * @param ypoints
     *        y-coordinates for the points that form the polygon.
     */
    public CCPolygon(double [] xpoints, double [] ypoints) {
        this(xpoints, ypoints, Color.black, Color.pink, new BasicStroke(1f));
    }
}

src/example/Main.java
package example;

import java.awt.Color;
import java.util.Random;

import javax.swing.JFrame;

import cartesian.coordinate.CCPoint;
import cartesian.coordinate.CCPolygon;
import cartesian.coordinate.CCSystem;
import cartesian.coordinate.CCLine;

public class Main extends JFrame {
    private static final long serialVersionUID = 1L;
    
    Main() {
        super("Viewer");
        setTitle("Viewer");
        
        setVisible(true);
        setSize(800, 600);
        setLocationRelativeTo(null);
        CCSystem s = new CCSystem(0.0, 0.0, 10.0, 10.0);
        add(s);
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        
        
        double[] x = new double[]{-1,2,0};
        double[] y = new double[]{-1,-2,3};
        CCPolygon ccp = new CCPolygon(x, y);
        
        s.add(new CCLine(1.0, 0.0, Color.blue));
        s.add(new CCLine(-1.0, 10.0, Color.magenta));
        s.add(new CCLine(1.0, 5.0, Color.red));
        s.add(new CCLine(-1.0, 5.0, Color.cyan));
        s.add(new CCLine(-1.0, 15.0, Color.yellow));
        s.add(new CCLine(1.0, -5.0, Color.green));
        s.add(new CCLine(1.0, 0.0, 5.0, Color.orange));
        s.add(new CCLine(0.0, 1.0, 5.0, Color.pink));
        s.add(new CCPoint(4, 5));
        s.add(ccp);
        
        /*
        Random r = new Random();
        for (int i = 0; i < 300; i++) {
            double a = r.nextDouble();
            double b = r.nextDouble();
            double c = r.nextDouble();
            if (r.nextBoolean()) a = -a;
            if (r.nextBoolean()) b = -b;
            if (r.nextBoolean()) c = -c;
            
            int cint = r.nextInt(13);
            Color color;
            
            switch (cint) {
            case 0: color = Color.black; break;
            case 1: color = Color.blue; break;
            case 2: color = Color.cyan; break;
            case 3: color = Color.darkGray; break;
            case 4: color = Color.gray; break;
            case 5: color = Color.green; break;
            case 6: color = Color.lightGray; break;
            case 7: color = Color.magenta; break;
            case 8: color = Color.orange; break;
            case 9: color = Color.pink; break;
            case 10: color = Color.red; break;
            case 11: color = Color.white; break;
            default: color = Color.yellow;
            }
            
            s.add(new CCLine(a,b,c,color));
        }
        */
    }
    public static void main(String[] args) {
        new Main();
    }
}
